sources/api/early-boot-config/src/provider/vmware.rs

```diff
@@ -0,0 +1,397 @@
+    // systemd unit file
+    const CD_ROM_MOUNT: &'static str = "/media/cdrom";
+    // A mounted CD-ROM may contain an OVF file or a user-supplied file named `user-data`
+    const USER_DATA_FILENAMES: [&'static str; 5] = [
+        "user-data",
+        "ovf-env.xml",
+        "OVF-ENV.XML",
+        "ovf_env.xml",
+        "OVF_ENV.XML",
+    ];
+
+    // The fields in which user data and its encoding are stored in guestinfo
+    const GUESTINFO_USERDATA: &'static str = "guestinfo.userdata";
+    const GUESTINFO_USERDATA_ENCODING: &'static str = "guestinfo.userdata.encoding";
+
+    /// Read and decode user data from files via mounted CD-ROM
+    fn cdrom_user_data() -> Result<Option<SettingsJson>> {
+        // Given the list of acceptable filenames, ensure only 1 exists and parse
+        // it for user data
+        info!("Attempting to retrieve user data from mounted CD-ROM");
+        let mut user_data_files = Self::USER_DATA_FILENAMES
+            .iter()
+            .map(|filename| Path::new(Self::CD_ROM_MOUNT).join(filename))
+            .filter(|file| file.exists());
+
+        let user_data_file = match user_data_files.next() {
+            Some(file) => file,
+            None => return Ok(None),
+        };
+
+        ensure!(
+            user_data_files.next().is_none(),
+            error::UserDataFileCount {
+                location: Self::CD_ROM_MOUNT
+            }
+        );
+
+        // XML files require extra processing, while a user-supplied file should already be in TOML
+        // format
+        info!("'{}' exists, using it", user_data_file.display());
+        let user_data_str = match user_data_file.extension().and_then(OsStr::to_str) {
+            Some("xml") | Some("XML") => Self::ovf_user_data(&user_data_file)?,
+            // Since we only look for a specific list of file names, we should never find a file
+            // with an extension we don't understand.
+            Some(_) => unreachable!(),
+            None => {
+                // Read the file, decompressing it if compressed.
+                expand_file_maybe(&user_data_file).context(error::InputFileRead {
+                    path: &user_data_file,
+                })?
+            }
+        };
+
+        if user_data_str.is_empty() {
+            return Ok(None);
+        }
+
+        // User data could be 700MB compressed!  Eek!  :)
+        if user_data_str.len() <= 2048 {
+            trace!("Received user data: {}", user_data_str);
+        } else {
+            trace!(
+                "Received long user data, starts with: {}",
+                // (this isn't perfect because chars aren't grapheme clusters, but will error
+                // toward printing the whole input, which is fine)
+                String::from_iter(user_data_str.chars().take(2048))
+            );
+        }
+
+        let json = SettingsJson::from_toml_str(&user_data_str, "user data from CD-ROM").context(
```

Comment:
```suggestion
        let json = SettingsJson::from_toml_str(&user_data_str, "user data from CD-ROM").with_context(||
```
nit: so we don't do the string conversion unnecessarily.